test/SlowTests/Cluster/ClusterOperationTests.cs

```diff
@@ -293,5 +298,403 @@ public static async Task ReverseOrderSuccessfully(IDocumentStore store, string d
+                        new QueryOperationOptions { AllowStale = false, StaleTimeout = TimeSpan.FromSeconds(30) }
+                    );
+                    
+                    var expectedNode = (await store.GetRequestExecutor().GetPreferredNode()).Node.ClusterTag;
+
+                    //set up waiting for changes api in WaitForCompletionAsync to set up web socket
+                    var database = await Databases.GetDocumentDatabaseInstanceFor(expectedNode, store.Database);
+                    var delayQueryByPatch = new AsyncManualResetEvent();
+                    database.ForTestingPurposesOnly().DelayQueryByPatch = delayQueryByPatch;
+                    
+                    var op = await store.Operations.SendAsync(patch, token: cts.Token);
+                    Assert.Equal(expectedNode, op.NodeTag);
+
+                    var t = op.WaitForCompletionAsync(cts.Token);
+                    
+                    var waitWebSocketError = new AsyncManualResetEvent(cts.Token);
+                    var changes = store.Changes(store.Database, op.NodeTag);
+                    changes.ConnectionStatusChanged += (sender, args) => { waitWebSocketError.Set(); };
+
+                    //wait for websocket to connect
+                    await AssertWaitForTrueAsync(() => Task.FromResult(changes.Connected));
+
+                    //bring down server
+                    var serverWithPatchOperation = clusterNodes.Single(x => x.ServerStore.NodeTag == expectedNode);
+                    var result = await DisposeServerAndWaitForFinishOfDisposalAsync(serverWithPatchOperation);
+
+                    //wait for websocket to throw and retry
+                    await waitWebSocketError.WaitAsync(cts.Token);
+                    
+                    //bring server back up
+                    var settings = new Dictionary<string, string> { { RavenConfiguration.GetKey(x => x.Core.ServerUrls), result.Url } };
+                    var server = GetNewServer(new ServerCreationOptions { RunInMemory = false, DeletePrevious = false, DataDirectory = result.DataDirectory, CustomSettings = settings, NodeTag = result.NodeTag });
+                    Servers.Add(server);
+
+                    //run the operation again - should work
+                    op = await store.Operations.SendAsync(patch, token: cts.Token);
+                    
+                    // we reconnect to the same node
+                    Assert.Equal(expectedNode, op.NodeTag);
+                    await op.WaitForCompletionAsync(cts.Token);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task ChangesApiForOperationShouldCleanUpFaultyConnection_AfterUnrecoverableError()
+        {
+            DoNotReuseServer();
+            using (var cts = new CancellationTokenSource(TimeSpan.FromMinutes(2)))
+            {
+                var (clusterNodes, leader) = await CreateRaftCluster(1, leaderIndex: 0, shouldRunInMemory: false);
```

Comment:
hmmm is there a reason to create a one node cluster?
I guess it is either that or `DoNotReuseServer`?

Reply:
Disposing the global server caused me some problems and I preferred not to touch it since others may be using it

Reply:
definitely :)